### PR TITLE
changes for randao

### PR DESCRIPTION
**Pull request by blockchaindevsh**
add `L1Block.blockHash` to fetch 8191 history L1 blockhash not including the current one.

TODO: 
1. add e2e test

### CODE REVIEW

File: packages/contracts-bedrock/src/L2/L1Block.sol [hunk 1]
```diff
@@ -115,5 +118,24 @@ contract L1Block is ISemver {
             sstore(hash.slot, calldataload(100)) // bytes32
             sstore(batcherHash.slot, calldataload(132)) // bytes32
         }
+
+        historyHashes[number % 257] = hash;
+    }
+
+    // translated from
+    // [opBlockhash](https://github.com/ethereum/go-ethereum/blob/e31709db6570e302557a9bccd681034ea0dcc246/core/vm/instructions.go#L434)
+    function blockHash(uint256 historyNumber) external view returns (bytes32) {
```
Comment (qizhou) [inline packages/contracts-bedrock/src/L2/L1Block.sol]: Could we have more standardized comments as in other methods?

**Review by qizhou — COMMENTED** [review]
Can we separate the two features (SoulETH and RANDAO) into two PRs?

File: packages/contracts-bedrock/src/L2/L1Block.sol (hunk 1 above)
Comment (blockchaindevsh) [reply packages/contracts-bedrock/src/L2/L1Block.sol]: Done.

**Comment by blockchaindevsh** [comment]
> Can we separate the two features (SoulETH and RANDAO) into two PRs?

SoulETH related changes are moved [here](https://github.com/ethstorage/optimism/pull/19).

ES part is moved [here](https://github.com/ethstorage/optimism/pull/20)

**Review by qizhou — APPROVED** [review]

**Review by qzhodl — APPROVED** [review]

File: packages/contracts-bedrock/src/L2/L1Block.sol [hunk 2]
```diff
@@ -115,5 +120,29 @@ contract L1Block is ISemver {
...
+        // [opBlockhash](https://github.com/ethereum/go-ethereum/blob/e31709db6570e302557a9bccd681034ea0dcc246/core/vm/instructions.go#L434)
+        // with 256 => HISTORY_SIZE-1
+        uint256 lower;
+        uint256 upper = number;
+        if (upper < HISTORY_SIZE) {
+            lower = 0;
+        } else {
+            lower = upper - HISTORY_SIZE + 1;
+        }
+        if (_historyNumber >= lower && _historyNumber < upper) {
```
Comment (qzhodl) [inline packages/contracts-bedrock/src/L2/L1Block.sol:142]: _historyNumber <= upper ?

**Pull request merged**